### PR TITLE
feat(extendRoutes): allow relative paths in EditableTreeNode

### PR DESCRIPTION
**Pull request by robertmoura**
BREAKING CHANGE: allow relative path overrides

Closes https://github.com/posva/unplugin-vue-router/issues/341

- Support relative paths for `EditableTreeNode`
- Require absolute paths at the base of the tree and the root's children

### CODE REVIEW

**Review by posva — CHANGES_REQUESTED** [review]
Could you add a test? I would like to see the error reproduced

**Comment by robertmoura** [comment]
@posva I'm not entirely sure how the trees are structured or created. I had to update the code to check that the parent was the root. Can you think of any cases where this would not work?

**Comment by posva** [comment]
Do you have a bug reproduction? This doesn't seem to be fixing something 😅 
Changing paths with non absolute routes is not supported yet #341

**Comment by robertmoura** [comment]
@posva Quite right! It's actually more of a feature 😊 https://github.com/posva/unplugin-vue-router/issues/341 is exactly the problem I was trying to solve 😊

**Comment by robertmoura** [comment]
I think I'm seeing the problem now. Here's a snippet out of `TreeNode`. Here the path is assumed to be relative (a path segment) or absolute if there is a path override. Interesting behaviour... There has to be a bit more thought put into this. Marking this as a draft 😊 

```
  get path() {
    return (
      this.value.overrides.path ??
      (this.parent?.isRoot() ? '/' : '') + this.value.pathSegment
    )
  }
```

**Comment by robertmoura** [comment]
If we were to make a slight change to the behaviour of the `TreeNode` I think relative paths could be supported. It would be a breaking change though. Can you see any reasons why this would be a bad idea?

File: src/core/extendRoutes.spec.ts [hunk 1]
```diff
@@ -251,4 +251,26 @@ describe('EditableTreeNode', () => {
...
+
+  it('can override the path', () => {
+    const tree = new PrefixTree(RESOLVED_OPTIONS)
+    const editable = new EditableTreeNode(tree)
+    const node = editable.insert('foo', 'file.vue')
+    const childNode = node.insert('bar', 'file.vue')
+
+    childNode.path = 'baz'
+    node.path = 'baz'
+    editable.path = 'baz'
```
Comment (posva) [inline src/core/extendRoutes.spec.ts:263]: FYI: the root node shouldn’t be changed. Or at least not for this

Comment (robertmoura) [reply src/core/extendRoutes.spec.ts:263]: Great! I'll update that now 😊

**Comment by posva** [comment]
The fullPath shouldn’t change its behavior, or at least it shouldn’t be needed

**Comment by robertmoura** [comment]
I'll have a look at where it is being used and see if I can remove it 😊

**Comment by robertmoura** [comment]
Sorry, I was overcomplicating this a lot. I didn't realise that you could have absolute paths as child routes. I just assumed you had to put an extra `/` in the URL or something. Does what I'm doing now make more sense?

File: src/core/extendRoutes.spec.ts [hunk 2]
```diff
@@ -251,4 +251,20 @@ describe('EditableTreeNode', () => {
...
+
+  it('can override the path', () => {
+    const tree = new PrefixTree(RESOLVED_OPTIONS)
+    const editable = new EditableTreeNode(tree)
+    const parent = editable.insert('parent', 'file.vue')
+    const child = parent.insert('child', 'file.vue')
+
+    child.path = 'relative'
+    parent.path = 'relative'
+    expect(parent.path).toBe('/parent')
```
Comment (posva) [inline src/core/extendRoutes.spec.ts:263]: shouldn't this be `relative`

File: src/core/extendRoutes.ts [hunk 3]
```diff
@@ -142,9 +142,12 @@ export class EditableTreeNode {
    * Override the path of the route. You must ensure `params` match with the existing path.
    */
   set path(path: string) {
-    if (!path.startsWith('/')) {
+    if (
+      (!this.node.parent || this.node.parent.isRoot()) &&
```
Comment (posva) [inline src/core/extendRoutes.ts:146]: the isRoot now checks for the parent (just pushed it)

File: src/core/extendRoutes.spec.ts (hunk 2 above)
Comment (robertmoura) [reply src/core/extendRoutes.spec.ts:264]: That was intentional to test the warnings. Is there a method you use to test warnings It should log this warning:

> Only absolute paths are supported at the root of the route tree. Make sure that "relative" starts with a slash "/".`

The idea is that you would only be restricted to absolute paths at the root and the root's children.

**Comment by posva** [comment]
Looking good. I will probably give this a review and merge next Monday during the stream

**Comment by robertmoura** [comment]
Closing in favour of https://github.com/posva/unplugin-vue-router/pull/519

**Pull request closed without merging**